### PR TITLE
Skip to next JPEG segment in case of corrupt header length

### PR DESCRIPTION
**Pull request by oliverguenther**
In case an image has an invalid header segment such as an incorrect
length, prawn will reject the image while other readers can account for this
error and still render the content fine.

JPEG segments must be `0xFF` followed by anything but `0x00`. `0xFF` are
escaped with `0xFF` in the header, so we can be sure that `0xFF0xXX` is
a segment control.

This picture and a whole bunch of others I got from an iPhone 7 camera raises a 'JPEG marker not found! error due to this line https://github.com/prawnpdf/prawn/blob/d980247be8a00e7c59cd4e5785e3aa98f9856db1/lib/prawn/images/jpg.rb#L46

20200203_100520114_iOS

### One-liner to reproduce

```ruby
Prawn::Images::JPG.new(
  open('https://user-images.githubusercontent.com/459462/74015137-5b0e3f00-4990-11ea-9b7d-0bc4290fb132.jpg').read
)

```

I printed the loop of headers it is going through. Looks like the specified length in the second segment is incorrect. I printed the marker locations until the error is raised:

```
[3] pry(main)> Prawn::Images::JPG.new(open('https://user-images.githubusercontent.com/459462/74015137-5b0e3f00-4990-11ea-9b7d-0bc4290fb132.jpg').read)
MARKER ff CODE e0 LENGTH 16
MARKER ff CODE e1 LENGTH 1916
MARKER 0 CODE 0 LENGTH 65499
RuntimeError: JPEG marker not found!
``` 

The second length appears to be 2 bytes short to the next  marker (FF DB).

This is also output by `jpeginfo` or imagemagick `identify`:

```bash

$ jpeginfo 20200203_100520114_iOS.jpg 
20200203_100520114_iOS.jpg  Corrupt JPEG data: 2 extraneous bytes before marker 0xdb 4032 x 3024 24bit Exif  N 4438201 

$ identify 20200203_100520114_iOS.jpg 
20200203_100520114_iOS.jpg JPEG 4032x3024+0+0 DirectClass 8-bit 4.2Mi 0.000u 0m:0.000010s
identify identify: Corrupt JPEG data: 2 extraneous bytes before marker 0xdb (20200203_100520114_iOS.jpg).
```

The image is still being loaded in many places correctly, so I think it's a good idea to make the header parsing a bit more robust, especially since no metadata other than dimensions and channels / bits are needed for prawn.

### CODE REVIEW

File: lib/prawn/images/jpg.rb [hunk 1]
```diff
@@ -43,7 +43,11 @@ def initialize(data)
         d.seek(2) # Skip the first two bytes of JPEG identifier.
         loop do
           marker, code, length = d.read(4).unpack('CCn')
-          raise 'JPEG marker not found!' if marker != c_marker
+
+          # Skip corrupted header lengths
+          while (marker != c_marker || code == 0x00)
+            marker, code, length = d.read(4).unpack('CCn')
+          end
```
Comment (pointlessone) [inline lib/prawn/images/jpg.rb:50]: This code doesn't quite work. It reads in chunks of 4 bytes. Your image has 2 junk bytes. So this code would overshoot and won't find any other chunks unless they happen to be 4-bytes aligned relative to the end of the last proper chunk. In the attached file that is not the case for any of the following chunks.

Another thing to consider is that `0xFF` can be used for padding between chunks. So the first `0xFF` might be not the actual marker. But this code would take it as one.

The correct way to handle this, I'm afraid, would be a bit more involved.

Here's an example of what I came up with that works for the attached file. I didn't thoroughly test it and would appreciate if you take a closer look at it:
```ruby
# Find marker, or rather the first instance of one
marker = 0 # Make sure it's not c_marker
marker = d.read(1).unpack1('C') until marker == c_marker

# Find code after all the padding marker bytes
code = marker 
code = d.read(1).unpack1('C') until code != c_marker

# Get length
length = d.read(2).unpack1('n')
```
This goes instead of lines 45—50.

Comment (petergoldstein) [reply lib/prawn/images/jpg.rb:50]: @pointlessone There's a higher order consideration here - is there a desire for prawn to be "flexible" in what it accepts in terms of JPG images?

Moreover, this contribution seems focused on one possible error out of many.  What JPG errors does prawn want to account for?  What errors does it fail on?

There are other tools that can be used to filter or correct bad JPGs should they occur in your workflow.  And in general, malformed JPGs are unusual - this is a very edgy case.

It's a very defensible position to say that prawn expects JPGs that are to be embedded in PDFs to be well-formed.  I would recommend that this be the position prawn takes.

Thoughts?

Comment (pointlessone) [reply lib/prawn/images/jpg.rb:50]: I think Prawn has to only accept valid JPEGs. We may provide some flexibility but I don't think it's useful. By accepting broken/invalid JPEGs we're pushing the issue down the line. We can't guaranty a document with a broken JPEG would be rendered by any renderer (e.g. Acrobat might swallow it but Preview or Evince might not).

We're also opening door to security exploits. How many times we've seen "specially crafted input" result in all sorts of nastiness. I don't think Prawn should be the guardian but we shouldn't spent time actively enabling it.

Prawn doesn't require much compliance anyway. JPEG is a natively supported blob in PDF. We don't process it in any way. It just needs to be good enough for us to extract a few facts about the image.

 TL;DR: We only support valid JPEGs. Fix your images elsewhere.

Comment (petergoldstein) [reply lib/prawn/images/jpg.rb:50]: @pointlessone Given that decision I think this PR should be closed.

**Comment by oliverguenther** [comment]
Thanks for your feedback and time to look into this :bow: 

Sorry this got a bit lost in my list of open things, and I agree that this is indeed an edge case that can be verified earlier on if necessary.
